### DISCUSSION TITLE
How to export Swagger to Yaml when accessed with docs.

### DISCUSSION THREAD

**Discussion by satodaiki** [fastapi/fastapi#6991] [Questions] [answered]
Hello.

As the title says, I always want to have the Swagger documentation that opens when I access `/ docs` as a yaml file, but I don't know how to do it.

I want to generate type information to be used in the front end from the output yaml file.

**Comment by insomnes** [comment]
@satodaiki Maybe it could help: there is OpenAPI specification at /openapi.json (also linked on top of /docs)

**Comment by satodaiki** [comment]
Thank you!

The way I came up with to save this json as a file every time I launch the fast api is to run a simple shell script like this after launch.

If there is, I'd like to use the json download method that the official follows, if any, do you know anything?

```
curl -O localhost:8000/openapi.json
```

**Comment by insomnes** [comment]
Glad that helped you! 
Can you please explain more, what do you mean by this:
> If there is, I'd like to use the json download method that the official follows, if any, do you know anything?

Now i can only say that by running:
`curl -O localhost:8000/openapi.json`
in some bash script you would save `openapi.json` file to script working dir.

**Comment by falkben** [comment]
`app.openapi()` will return the openapi schema as a dictionary. 

you could convert this to yaml file, and serve it on /docs if you wanted...

**Comment by satodaiki** [comment]
I'm sorry, I was resting😪

> Glad that helped you!
> Can you please explain more, what do you mean by this:
> 
> > If there is, I'd like to use the json download method that the official follows, if any, do you know anything?
> 
> Now i can only say that by running:
> `curl -O localhost:8000/openapi.json`
> in some bash script you would save `openapi.json` file to script working dir.

I will answer about this.
Now I'm running a simple script like this to launch the fastapi app.

```
uvicorn app.main: app --reload
```

If the official provided a way to download openapi.json to the root directory every time I ran this command and launched the app, I wanted to know how!

**Comment by insomnes** [comment]
@satodaiki as @falkben mentioned you can get openapi data in app, so you can run [startup function](https://fastapi.tiangolo.com/advanced/events/) like this:
```
import json

from fastapi import FastAPI


app = FastAPI()


@app.on_event("startup")
def save_openapi_json():
    openapi_data = app.openapi()
    # Change "openapi.json" to desired filename
    with open("openapi.json", "w") as file:
        json.dump(openapi_data, file)
```
But there is catch if you run uvicorn as:
`uvicorn app.main: app --reload`
It would reload permanently, cause file would be overwritten in this simple example. You can check for changes before writing or remove `--reload` option

**Comment by satodaiki** [comment]
@insomnes 
Thank you!
Now that the issue is resolved, close the issue.

**Comment by tiangolo** [comment]
Thanks for the help here @insomnes and @falkben ! 👏  🙇 

Thanks for reporting back and closing the issue @satodaiki  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by gamgi** [comment]
I also wrote an [article on how to export OpenAPI spec from FastAPI to json or yaml](https://www.doctave.com/blog/python-export-fastapi-openapi-spec). If you're doing this from CI, for example for code generation, using `uvicorn.importer` is slightly more convenient.

**Comment by glowinthedark** [comment, marked as answer]
the easiest way — add an endpoint to download your swagger as YAML e.g. via http://localhost:8000/yaml

```python
import yaml
import uvicorn

from fastapi import FastAPI
from fastapi.responses import Response

app = FastAPI(description="FastAPI App", debug=True)

# 
# ... your model definitions/imports and endpoints go here
#

@app.get("/yaml", response_class=Response, include_in_schema=False)
def get_openapi_yaml():
    openapi_spec = app.openapi()
    openapi_yaml = yaml.dump(openapi_spec, default_flow_style=False)

    return Response(content=openapi_yaml, media_type="application/x-yaml")

uvicorn.run(app, host="0.0.0.0", port=8000)
```

> <sup>`include_in_schema=False` implies you do NOT want this specific endpoint to be part of your swagger schema</sup>
